docs/core/extensions/async-state.md

```diff
@@ -0,0 +1,290 @@
+{
+    public string CorrelationId { get; set; } = string.Empty;
+}
+```
+
+In this example, the correlation ID is set once at the beginning of request processing and is automatically available in all subsequent async operations without needing to pass it as a parameter.
+
+## ASP.NET Core integration
+
+In ASP.NET Core applications, you can use async state to flow request-specific information through your application:
+
+```csharp
+using Microsoft.AspNetCore.Builder;
+using Microsoft.Extensions.AsyncState;
+using Microsoft.Extensions.DependencyInjection;
+
+var builder = WebApplication.CreateBuilder(args);
+
+builder.Services.AddAsyncState();
+builder.Services.AddScoped<RequestHandler>();
+
+var app = builder.Build();
+
+app.Use(async (context, next) =>
+{
+    var asyncContext = context.RequestServices.GetRequiredService<IAsyncContext<RequestMetadata>>();
+    
+    // Set request metadata at the beginning of the pipeline
+    asyncContext.Set(new RequestMetadata
+    {
+        RequestId = context.TraceIdentifier,
+        RequestPath = context.Request.Path,
+        StartTime = DateTimeOffset.UtcNow
+    });
+
+    await next();
+});
+
+app.MapGet("/api/data", async (RequestHandler handler) =>
+{
+    return await handler.GetDataAsync();
+});
+
+app.Run();
+
+public class RequestHandler
+{
+    private readonly IAsyncContext<RequestMetadata> _asyncContext;
+
+    public RequestHandler(IAsyncContext<RequestMetadata> asyncContext)
+    {
+        _asyncContext = asyncContext;
+    }
+
+    public async Task<object> GetDataAsync()
+    {
+        await Task.Delay(50);
+        
+        if (_asyncContext.TryGet(out var metadata))
+        {
+            var duration = DateTimeOffset.UtcNow - metadata.StartTime;
+            return new
+            {
+                RequestId = metadata.RequestId,
+                RequestPath = metadata.RequestPath,
+                Duration = duration.TotalMilliseconds
+            };
+        }
+
+        return new { Error = "No request metadata available" };
+    }
+}
+
+public class RequestMetadata
+{
+    public string RequestId { get; set; } = string.Empty;
+    public string RequestPath { get; set; } = string.Empty;
+    public DateTimeOffset StartTime { get; set; }
+}
+```
+
+## Best practices
+
+When using async state, consider the following best practices:
+
+- **Limit state size**: Keep async state objects small to minimize memory overhead and serialization costs in scenarios where async state might need to be serialized.
+- **Initialize state early**: Set async state values as early as possible in your async operation to ensure they're available throughout the execution flow.
+- **Clean up state**: Reset or clear async state when it's no longer needed to avoid memory leaks in long-running applications.
+- **Use appropriate interfaces**: Use `IAsyncContext<T>` when you need explicit control over state initialization and reset. Use `IAsyncState` for simpler property-based access.
```

Comment:
That's backwards.  IAsyncState is where you control the lifetime (e.g. per request).